core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java

```diff
@@ -60,25 +59,26 @@
    @Inject private CommandsFactory factory;
    @Inject private IteratorHandler iteratorHandler;
    private boolean hasLoader;
+   private int maxSegment;
 
    private Address localAddress;
 
    private final ConcurrentMap<Object, SegmentListener> changeListener = CollectionFactory.makeConcurrentMap();
    private ByteString cacheName;
 
    class SegmentListener {
-      private final Set<Integer> segments;
+      private final IntSet segments;
       private final SegmentAwareOperation op;
-      private final Set<Integer> segmentsLost;
+      private final IntSet segmentsLost;
 
-      SegmentListener(Set<Integer> segments, SegmentAwareOperation op) {
-         this.segments = new HashSet<>(segments);
+      SegmentListener(IntSet segments, SegmentAwareOperation op) {
+         this.segments = segments;
          this.op = op;
-         this.segmentsLost = Collections.synchronizedSet(new HashSet<>());
+         this.segmentsLost = IntSets.concurrentSet(maxSegment);
       }
 
-      public void localSegments(Set<Integer> localSegments) {
-         segments.forEach(s -> {
+      public void localSegments(IntSet localSegments) {
+         segments.forEach((int s) -> {
```

Comment:
I should say something about lambdas again, after all `lostSegments()` already uses an enhanced for loop :)

Reply:
Okay